File: crates/brainwires-mcp-server/examples/mcp_server.rs
//! MCP Server with tool registry and middleware pipeline.
//!
//! Demonstrates standing up an `McpServer` with:
//! - A `McpToolRegistry` populated with `McpToolDef` entries
//! - A middleware chain combining auth, logging, rate-limiting, and tool filtering
//! - Server lifecycle (build, configure, summarize)
//!
//! The example does **not** actually start the event loop (that requires a
//! real transport and a connected MCP client), but it exercises the full
//! construction API so you can see how the pieces fit together.
//!
//! ```bash
//! cargo run -p brainwires-network --example mcp_server --features server
//! ```

use anyhow::Result;
use async_trait::async_trait;
use brainwires_mcp_client::{CallToolResult, ServerCapabilities, ServerInfo};
use brainwires_mcp_server::{
    AuthMiddleware, LoggingMiddleware, McpServer, McpToolDef, McpToolRegistry, MiddlewareChain,
    RateLimitMiddleware, RequestContext, ToolFilterMiddleware, ToolHandler,
};
use serde_json::{Value, json};

// ---------------------------------------------------------------------------
// 1. Define tool handlers
// ---------------------------------------------------------------------------

/// A simple tool handler that echoes back the input arguments.
struct EchoHandler;

#[async_trait]
impl ToolHandler for EchoHandler {
    async fn call(&self, args: Value, _ctx: &RequestContext) -> Result<CallToolResult> {
        println!("    [EchoHandler] called with: {args}");
        Ok(CallToolResult::success(vec![]))
    }
}

/// A tool handler that returns the current (simulated) time.
struct TimeHandler;

#[async_trait]
impl ToolHandler for TimeHandler {
    async fn call(&self, _args: Value, _ctx: &RequestContext) -> Result<CallToolResult> {
        let now = chrono::Utc::now().to_rfc3339();
        println!("    [TimeHandler] returning time: {now}");
        Ok(CallToolResult::success(vec![]))
    }
}

// ---------------------------------------------------------------------------
// 2. Define an McpHandler backed by the tool registry
// ---------------------------------------------------------------------------

struct DemoHandler {
    registry: McpToolRegistry,
}

#[async_trait]
impl brainwires_mcp_server::McpHandler for DemoHandler {
    fn server_info(&self) -> ServerInfo {
        ServerInfo {
            name: "demo-mcp-server".to_string(),
            version: "0.7.0".to_string(),
        }
    }

    fn capabilities(&self) -> ServerCapabilities {
        ServerCapabilities::default()
    }

    fn list_tools(&self) -> Vec<McpToolDef> {
        self.registry.list_tool_defs()
    }

    async fn call_tool(
        &self,
        name: &str,
        args: Value,
        ctx: &RequestContext,
    ) -> Result<CallToolResult> {
        self.registry.dispatch(name, args, ctx).await
    }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

#[tokio::main]
async fn main() -> Result<()> {
    println!("=== MCP Server Example ===\n");

    // Step 1: Build the tool registry
    println!("--- Tool Registry ---");

    let mut registry = McpToolRegistry::new();

    registry.register(
        "echo",
        "Echoes back the provided arguments",
        json!({
            "type": "object",
            "properties": {
                "message": { "type": "string", "description": "The message to echo" }
            },
            "required": ["message"]
        }),
        EchoHandler,
    );

    registry.register(
        "get_time",
        "Returns the current UTC time",
        json!({ "type": "object", "properties": {} }),
        TimeHandler,
    );

    for def in registry.list_tools() {
        println!("  Registered tool: {} — {}", def.name, def.description);
    }
    println!();

    // Step 2: Build the middleware chain
    println!("--- Middleware Chain ---");

    let mut chain = MiddlewareChain::new();

    // Auth: require a bearer token for non-initialize requests
    chain.add(AuthMiddleware::new("demo-secret-token"));
    println!("  1. AuthMiddleware       (token = demo-secret-token)");

    // Logging: log every request
    chain.add(LoggingMiddleware::new());
    println!("  2. LoggingMiddleware");

    // Rate limit: 20 requests per second
    chain.add(RateLimitMiddleware::new(20.0));
    println!("  3. RateLimitMiddleware  (20 req/s)");

    // Tool filter: block dangerous tools
    chain.add(ToolFilterMiddleware::deny(["rm_rf", "drop_database"]));
    println!("  4. ToolFilterMiddleware (deny: rm_rf, drop_database)");
    println!();

    // Step 3: Verify the registry can dispatch
    println!("--- Dispatch Test ---");

    let ctx = RequestContext::new(json!(1));

    println!("  Dispatching 'echo':");
    let result = registry
        .dispatch("echo", json!({"message": "hello"}), &ctx)
        .await;
    match &result {
        Ok(_) => println!("    Result: OK"),
        Err(e) => println!("    Result: Error — {e}"),
    }

    println!("  Dispatching 'get_time':");
    let result = registry.dispatch("get_time", json!({}), &ctx).await;
    match &result {
        Ok(_) => println!("    Result: OK"),
        Err(e) => println!("    Result: Error — {e}"),
    }

    println!("  Dispatching 'nonexistent_tool':");
    let result = registry.dispatch("nonexistent_tool", json!({}), &ctx).await;
    match &result {
        Ok(_) => println!("    Result: OK (unexpected)"),
        Err(e) => println!("    Result: Error — {e}"),
    }
    println!();

    // Step 4: Show how McpServer is constructed (without running the event loop)
    println!("--- Server Construction ---");

    let handler = DemoHandler { registry };
    let server = McpServer::new(handler)
        .with_middleware(AuthMiddleware::new("secret"))
        .with_middleware(LoggingMiddleware::new())
        .with_middleware(RateLimitMiddleware::new(10.0))
        .with_middleware(ToolFilterMiddleware::deny(["rm_rf"]));

    println!("  McpServer created with handler + 4 middleware layers");
    println!("  (Skipping server.run() — requires a connected transport)");

    // Prevent unused-variable warning
    drop(server);

    println!("\nDone.");
    Ok(())
}

File: deno/examples/network/mcp_server.ts
// Example: MCP Server — tool registry, middleware pipeline, and server construction
// Demonstrates building an McpToolRegistry with tool handlers, assembling a
// MiddlewareChain (auth, logging, rate-limiting, tool filtering), dispatching
// tool calls, and constructing an McpServer with middleware layers.
// Run: deno run deno/examples/agent-network/mcp_server.ts

import {
  AuthMiddleware,
  LoggingMiddleware,
  type McpHandler,
  McpServer,
  type McpToolDef,
  McpToolRegistry,
  MiddlewareChain,
  RateLimitMiddleware,
  RequestContext,
  ToolFilterMiddleware,
  type ToolHandler,
} from "@rullama/network";

import type {
  CallToolResult,
  ServerCapabilities,
  ServerInfo,
} from "@rullama/mcp-client";

async function main(): Promise<void> {
  console.log("=== MCP Server Example ===\n");

  // ---------------------------------------------------------------------------
  // 1. Build the tool registry with handlers
  // ---------------------------------------------------------------------------
  console.log("--- Tool Registry ---");

  const registry = new McpToolRegistry();

  // Echo handler: returns the input arguments as text
  const echoHandler: ToolHandler = async (
    args: Record<string, unknown>,
    _ctx: RequestContext,
  ): Promise<CallToolResult> => {
    const message = (args.message as string) ?? "(empty)";
    console.log(`    [EchoHandler] called with: ${message}`);
    return {
      content: [{ type: "text", text: `Echo: ${message}` }],
    };
  };

  registry.register(
    "echo",
    "Echoes back the provided arguments",
    {
      type: "object",
      properties: {
        message: { type: "string", description: "The message to echo" },
      },
      required: ["message"],
    },
    echoHandler,
  );

  // Time handler: returns the current UTC time
  const timeHandler: ToolHandler = async (
    _args: Record<string, unknown>,
    _ctx: RequestContext,
  ): Promise<CallToolResult> => {
    const now = new Date().toISOString();
    console.log(`    [TimeHandler] returning time: ${now}`);
    return {
      content: [{ type: "text", text: now }],
    };
  };

  registry.register(
    "get_time",
    "Returns the current UTC time",
    { type: "object", properties: {} },
    timeHandler,
  );

  for (const def of registry.listTools()) {
    console.log(`  Registered tool: ${def.name} — ${def.description}`);
  }
  console.log(`  Has "echo": ${registry.hasTool("echo")}`);
  console.log(`  Has "missing": ${registry.hasTool("missing")}`);
  console.log();

  // ---------------------------------------------------------------------------
  // 2. Build the middleware chain
  // ---------------------------------------------------------------------------
  console.log("--- Middleware Chain ---");

  const chain = new MiddlewareChain();

  // Auth: require a bearer token for non-initialize requests
  chain.add(new AuthMiddleware("demo-secret-token"));
  console.log("  1. AuthMiddleware       (token = demo-secret-token)");

  // Logging: log every request
  chain.add(new LoggingMiddleware());
  console.log("  2. LoggingMiddleware");

  // Rate limit: 20 requests per second
  chain.add(new RateLimitMiddleware(20));
  console.log("  3. RateLimitMiddleware  (20 req/s)");

  // Tool filter: block dangerous tools
  chain.add(ToolFilterMiddleware.deny(["rm_rf", "drop_database"]));
  console.log("  4. ToolFilterMiddleware (deny: rm_rf, drop_database)");
  console.log();

  // ---------------------------------------------------------------------------
  // 3. Dispatch tool calls through the registry
  // ---------------------------------------------------------------------------
  console.log("--- Dispatch Test ---");

  const ctx = new RequestContext(1);

  console.log("  Dispatching 'echo':");
  try {
    const result = await registry.dispatch("echo", { message: "hello" }, ctx);
    console.log(
      `    Result: OK — ${
        result.content[0]?.type === "text" ? result.content[0].text : "?"
      }`,
    );
  } catch (e) {
    console.log(`    Result: Error — ${e}`);
  }

  console.log("  Dispatching 'get_time':");
  try {
    const result = await registry.dispatch("get_time", {}, ctx);
    console.log(
      `    Result: OK — ${
        result.content[0]?.type === "text" ? result.content[0].text : "?"
      }`,
    );
  } catch (e) {
    console.log(`    Result: Error — ${e}`);
  }

  console.log("  Dispatching 'nonexistent_tool':");
  try {
    await registry.dispatch("nonexistent_tool", {}, ctx);
    console.log("    Result: OK (unexpected)");
  } catch (e) {
    console.log(`    Result: Error — ${e}`);
  }
  console.log();

  // ---------------------------------------------------------------------------
  // 4. Process requests through the middleware chain
  // ---------------------------------------------------------------------------
  console.log("--- Middleware Processing ---");

  // Simulate an initialize request (should pass auth)
  const initRequest = {
    jsonrpc: "2.0" as const,
    id: 1,
    method: "initialize",
    params: undefined,
  };
  const initCtx = new RequestContext(1);
  const initResult = await chain.processRequest(initRequest, initCtx);
  console.log(
    `  initialize -> ${
      initResult ? `Rejected: ${initResult.message}` : "Continue"
    }`,
  );

  // Simulate a tools/call request (will be rejected by auth — no token)
  const toolRequest = {
    jsonrpc: "2.0" as const,
    id: 2,
    method: "tools/call",
    params: { name: "echo", arguments: { message: "test" } },
  };
  const toolCtx = new RequestContext(2);
  const toolResult = await chain.processRequest(toolRequest, toolCtx);
  console.log(
    `  tools/call 'echo' (no token) -> ${
      toolResult ? `Rejected: ${toolResult.message}` : "Continue"
    }`,
  );

  // Simulate a tools/call with valid auth token
  const authToolRequest = {
    jsonrpc: "2.0" as const,
    id: 3,
    method: "tools/call",
    params: {
      name: "echo",
      arguments: { message: "test" },
      _auth_token: "demo-secret-token",
    },
  };
  const authToolCtx = new RequestContext(3);
  const authToolResult = await chain.processRequest(
    authToolRequest,
    authToolCtx,
  );
  console.log(
    `  tools/call 'echo' (with token) -> ${
      authToolResult ? `Rejected: ${authToolResult.message}` : "Continue"
    }`,
  );
  console.log();

  // ---------------------------------------------------------------------------
  // 5. Construct an McpServer (without running the event loop)
  // ---------------------------------------------------------------------------
  console.log("--- Server Construction ---");

  // Build a handler that delegates to the registry
  const handler: McpHandler = {
    serverInfo(): ServerInfo {
      return { name: "demo-mcp-server", version: "0.1.0" };
    },
    capabilities(): ServerCapabilities {
      return {};
    },
    listTools(): McpToolDef[] {
      return registry.listTools();
    },
    async callTool(
      name: string,
      args: Record<string, unknown>,
      reqCtx: RequestContext,
    ): Promise<CallToolResult> {
      return registry.dispatch(name, args, reqCtx);
    },
  };

  const server = new McpServer(handler)
    .withMiddleware(new AuthMiddleware("secret"))
    .withMiddleware(new LoggingMiddleware())
    .withMiddleware(new RateLimitMiddleware(10))
    .withMiddleware(ToolFilterMiddleware.deny(["rm_rf"]));

  console.log("  McpServer created with handler + 4 middleware layers");
  console.log("  Server info:", JSON.stringify(handler.serverInfo()));
  console.log("  (Skipping server.run() — requires a connected transport)");

  // Prevent unused-variable lint
  void server;

  console.log("\nDone.");
}

await main();
